In [1]:
# 导入必要的库
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

In [2]:
# 加载数据
bc_data = pd.read_csv('bc_data.csv', header=0)
data = bc_data.drop(['id'], axis=1)
X_data = data.drop(['diagnosis'], axis=1)
y_data = np.ravel(data[['diagnosis']])

In [3]:
# 数据标准化
means_norm = StandardScaler()
means_norm.fit(X_data)
X_data = means_norm.transform(X_data)

In [4]:
# 划分数据集
X_trainSet, X_testSet, y_trainSet, y_testSet = train_test_split(X_data, y_data, test_size=0.25, random_state=1)

In [5]:
# # 创建SVM模型 with default regularization
# svm_model = SVC(kernel='linear')

In [13]:
# # 创建SVM模型 with C=0.2 regularization
# svm_model = SVC(kernel='linear', C=0.2)

In [39]:
# # SVM with polynomial kernel
# svm_model = SVC(kernel='poly', C=0.2)

# # SVM with RBF kernel
# svm_model = SVC(kernel='rbf', C=0.2)

# SVM with sigmoid kernel
svm_model = SVC(kernel='sigmoid', C=0.2)

In [40]:
# 训练模型
svm_model.fit(X_trainSet, y_trainSet)

SVC(C=0.2, kernel='sigmoid')

In [41]:
# 预测测试集
y_pred_svm = svm_model.predict(X_testSet)

In [42]:
# 计算准确率
accuracy_svm = accuracy_score(y_testSet, y_pred_svm)
print(accuracy_svm)

0.958041958041958


> linear kernel
- 0.972027972027972 with default(c=1) regularization
- 0.958041958041958 with c=0.2 regularization

> polynomial kernel
- 0.8881118881118881 with default(c=1) regularization
- 0.8391608391608392 with c=0.2 regularization

> rbf kernel
- 0.965034965034965 with default(c=1) regularization
- 0.951048951048951 with c=0.2 regularization

> sigmoid kernel
- 0.951048951048951 with default(c=1) regularization
- 0.958041958041958 with c=0.2 regularization